numpy/core/src/umath/loops.c.src

```diff
@@ -2195,14 +2190,11 @@ HALF_@kind@(char **args, npy_intp *dimensions, npy_intp *steps, void *NPY_UNUSED
 {
     /*  */
     BINARY_LOOP {
-        npy_half in1 = *(npy_half *)ip1;
+        const npy_half in1 = *(npy_half *)ip1;
         const npy_half in2 = *(npy_half *)ip2;
-        in1 = (@OP@(in1, in2) || npy_half_isnan(in1)) ? in1 : in2;
-        if (npy_half_isnan(in1)) {
-            npy_set_floatstatus_invalid();
-        } 
-        *((npy_half *)op1) = in1;
+        *((npy_half *)op1) = (@OP@(in1, in2) || npy_half_isnan(in1)) ? in1 : in2;
     }
+    /* npy_half_isnan will never set floatstatus_invalid, so do not clear */
```

Comment:
when *do* we set floatstatus_invalid? is that set at the C level?

Reply:
Yes, the non-SSE code sets it in `a > b` comparisons. which is why I need to clear it in the float-type loops at line 1849, and the complex-type loops at line 2763,

Reply:
Well, set at the FPU level, not at the C level.

Can't we just reorder the comparisons to do the nan checks first, rather than messing with the FPU flags?

Reply:
trying, let's see what CI finds

Reply:
It turns out simd code also sets it, in a way that it is impossible to avoid the clear after the loop finishes.

Reply:
Rearranging the order still sets the flag since only one value is checked. It seems cheaper to me to clear the flag than to check both values. In changeset 46452b360 I added back the call to clear the flags, now tests are passing.

Reply:
I don't think instruction ordering is safe with modern compilers, at least without some work to make the order stick.